📝 Add issue and pull request templates

amulware opened the pull request:
## ✨ What's this?
This PR adds one pull request template (which you see in action here) and several issue templates to help us and others create well-formatted PRs and issues.

### 🔗 Relationships

Part of #176

## 🔍 Why do we want this?
To make it easier for us and others to submit issues and PRs in a well-formatted way. Generally, these sort of templates makes people think a bit more about what they write and help them add more relevant information than they otherwise would which is good for everyone!

## 🏗 How is it done?
By adding markdown files to the `.github` directory. Specifically by adding one pull request template, and several issue templates. When creating a new issue after this is merged, one will be prompted to choose from one of the issue templates which will populate the description and labels of the issue, and all PRs created here on GitHub will automatically have the PR template pre-filled.

All those files also include a bunch of HTML comments which explain each section and what to add there. Ideally, people will replace them with actual content, but if they forget, they won't be rendered in the final issue/PR, so at least we don't have to look at them everywhere (though they'll still show up in our webhook integrations probably..).

### 💥 Breaking changes
Absolutely none. :)

### 🔬 Why not another way?
We could also add multiple pull request templates (it's supported), however, there is no UI to chose from them, and we would have to add `?template=template-file.md` to url manually to use them, which isn't a great solution. I think sticking to a more general PR template is the better tradeoff here by far.

### 🦋 Side effects
Creating issues and PRs will take more time. However, they will also be much easier to read. Worth it!

If people forget to remove the HTML comments, our webhook integration posts will be ugly. However, I think there's too much value in the comments to remove them.

## 💡 Review hints
Read the markdown files, see if you like the layout of this PR, and if you're not sure how the issue templates are gonna look like, check a repo that supports them, or take a look at the documentation here: https://help.github.com/en/articles/about-issue-and-pull-request-templates

tomrijnbeek commented:
I have no immediate concerns with this PR. I am however curious as to the source of the template. Right now what is missing from the PR description is why the templates were chosen as they are (which isn’t trivial), e.g. whether they were based on an example. Not that I see anything wrong with them, but I wonder if there’s any best practices around these that we should try to make sure to apply.

amulware commented:
The templates are based on my experience of what works/is useful, filtered by what I could come up with/remember at the moment.

tomrijnbeek commented on .github/ISSUE_TEMPLATE/b-bug-report.md:
```diff
@@ -0,0 +1,23 @@
+<!-- Please give the issue a short, clear title -->
+
+## 🐛 Bug Report
+<!-- Summarise the bug here as clearly and concisely as you can -->
```
nit: leave empty line below markdown header. You do this for the question template

amulware replied:
I removed the empty line in the question template since it was the only place that had it. I see no reason for needing it.

tomrijnbeek replied:
Markdownlint [1] suggests blank lines around the header on both sides (which is what I personally use myself), but doesn't provide a particularly strong argument. I would prefer the blank lines, but this is mostly for aesthetic reasons, so it is not a super strong opinion.

[1] https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md#md022---headings-should-be-surrounded-by-blank-lines

amulware replied:
Under normal circumstances, I would empty lines but I think in the case of these comments it's fine. (I would also leave two empty lines above ## (h2) for better separation, but I don't think that does any good here.

tomrijnbeek commented on .github/ISSUE_TEMPLATE/b-bug-report.md:
```diff
@@ -0,0 +1,23 @@
+
+## 🐛 Bug Report
+<!-- Summarise the bug here as clearly and concisely as you can -->
+<!-- If you know where the bug is located, include links to the respective files or even lines of code -->
```
We should also ask for a stack trace if relevant.

tomrijnbeek commented on .github/ISSUE_TEMPLATE/c-feature-request.md:
```diff
@@ -0,0 +1,16 @@
+<!-- Please give the issue a short, clear title -->
+
+## ✨ Feature Request
+<!-- Summarise the feature you would like to see here as clearly and concisely as you can -->
```
very nit: you don’t say “here” in the other descriptions

tomrijnbeek commented on .github/ISSUE_TEMPLATE/c-feature-request.md:
```diff
@@ -0,0 +1,16 @@
+<!-- Summarise the feature you would like to see here as clearly and concisely as you can -->
+
+## Context
+<!-- Tell us what this feature would be useful for; who would use it, for what purpose? -->
```
very nit: incorrect use of semicolon. Should be a colon or full stop.

tomrijnbeek commented on .github/ISSUE_TEMPLATE/c-feature-request.md:
```diff
@@ -0,0 +1,16 @@
+<!-- Tell us what this feature would be useful for; who would use it, for what purpose? -->
+
+## Scope
+<!-- List various pieces of functionality you would like this feature to have. This could be a list of concrete functions, or just some bullet points explaining the various things that can be done with it -->
```
Do we also ask to mention things specifically out of scope?

tomrijnbeek commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,29 @@
+## 🏗 How is it done?
+<!-- Describe how the changes are implemented, list the different parts the changes consist of if it's more than one -->
+
+### 💥 Breaking changes
```
While it’s good to see breaking changes documented as part of a PR, I actually think it would be better documented in a changelog inside the repository (would also be visible in the review), since breaking changes should be documented for releases. See #173

amulware replied:
No harm in having this here, and who knows when the other thing gets implemented consistently.

tomrijnbeek commented:
General comment about this PR: the issue template folder is allcaps, but the pull request folder is not. Is this intentional?

amulware commented:
RE capitalisation: I believe the all-caps folder name to be needed (it may not be), and there is no pull request folder.

Addressed all comments, good for re-review afaik.
